File: novascript/src/frontend/ast.ts

// src/frontend/ast.ts

export type NodeType = 
  | "Program"
  | "VarDeclaration"
  | "GlobalDeclaration"
  | "FunctionDeclaration"
  | "IfStatement"
  | "WhileStatement"
  | "ForStatement"
  | "SwitchStatement"
  | "CaseStatement"
  | "TryCatchStatement"
  | "ReturnStatement"
  | "ThrowStatement"
  | "BreakStatement"
  | "ContinueStatement"
  | "ImportStatement"
  | "ImportExpr"
  | "AssignmentExpr"
  | "BinaryExpr"
  | "CallExpr"
  | "MemberExpr"
  | "AwaitExpr"
  | "Identifier"
  | "NumericLiteral"
  | "StringLiteral"
  | "ObjectLiteral"
  | "Property"
  | "ArrayLiteral"
  | "ArrowFnExpr"
  | "ExportDeclaration"
  | "NamedImportStatement";

export interface Statement {
  kind: NodeType;
  line?: number;
  column?: number;
  file?: string;
}

export interface Program extends Statement {
  kind: "Program";
  body: Statement[];
}

export interface VarDeclaration extends Statement {
  kind: "VarDeclaration";
  constant: boolean;
  identifier: string;
  value?: Expression;
}

export interface GlobalDeclaration extends Statement {
  kind: "GlobalDeclaration";
  constant: boolean;
  identifier: string;
  value?: Expression;
}

export interface FunctionDeclaration extends Statement {
  kind: "FunctionDeclaration";
  name: string;
  parameters: string[];
  body: Statement[];
  async: boolean;
}

export interface IfStatement extends Statement {
  kind: "IfStatement";
  condition: Expression;
  thenBranch: Statement[];
  elseBranch?: Statement[];
}

export interface WhileStatement extends Statement {
  kind: "WhileStatement";
  condition: Expression;
  body: Statement[];
}

export interface ForStatement extends Statement {
    kind: "ForStatement";
    // Nova-style
    counter?: string;
    start?: Expression;
    end?: Expression;
    // JS-style
    init?: Statement;
    condition?: Expression;
    update?: Expression;
    
    body: Statement[];
}

export interface SwitchStatement extends Statement {
    kind: "SwitchStatement";
    discriminant: Expression;
    cases: CaseStatement[];
    default?: Statement[];
}

export interface CaseStatement extends Statement {
    kind: "CaseStatement";
    test: Expression;
    consequent: Statement[];
}

export interface TryCatchStatement extends Statement {
    kind: "TryCatchStatement";
    body: Statement[];
    catchParameter?: string;
    catchBlock: Statement[];
    finallyBlock?: Statement[];
}

export interface ReturnStatement extends Statement {
  kind: "ReturnStatement";
  value?: Expression;
}

export interface ThrowStatement extends Statement {
    kind: "ThrowStatement";
    argument: Expression;
}

export interface BreakStatement extends Statement {
    kind: "BreakStatement";
}

export interface ContinueStatement extends Statement {
    kind: "ContinueStatement";
}

export interface ImportStatement extends Statement {
    kind: "ImportStatement";
    moduleName: string;
}

export interface NamedImportStatement extends Statement {
    kind: "NamedImportStatement";
    imports: string[];     // names to import, e.g. ["x", "y"]
    moduleName: string;    // the source module
}

export interface ImportExpr extends Expression {
    kind: "ImportExpr";
    moduleName: Expression;
}

export interface ExportDeclaration extends Statement {
    kind: "ExportDeclaration";
    declaration: Statement; // the var/const/fn being exported
}

export interface Expression extends Statement {}

export interface AssignmentExpr extends Expression {
  kind: "AssignmentExpr";
  assignee: Expression; // Usually Identifier or MemberExpr
  value: Expression;
}

export interface BinaryExpr extends Expression {
  kind: "BinaryExpr";
  left: Expression;
  right: Expression;
  operator: string;
}

export interface CallExpr extends Expression {
  kind: "CallExpr";
  args: Expression[];
  caller: Expression;
}

export interface MemberExpr extends Expression {
  kind: "MemberExpr";
  object: Expression;
  property: Expression; // Identifier or String
  computed: boolean; // if true, object[property]. if false, object.property
  optional: boolean; // if true, object?.property
}

export interface AwaitExpr extends Expression {
    kind: "AwaitExpr";
    argument: Expression;
}

export interface Identifier extends Expression {
  kind: "Identifier";
  symbol: string;
}

export interface NumericLiteral extends Expression {
  kind: "NumericLiteral";
  value: number;
}

export interface StringLiteral extends Expression {
  kind: "StringLiteral";
  value: string;
}

export interface Property extends Expression {
  kind: "Property";
  key: string;
  value?: Expression; // Optional because { key } shorthand could exist, or `name "Alice"` style
}

export interface ObjectLiteral extends Expression {
  kind: "ObjectLiteral";
  properties: Property[];
}

export interface ArrayLiteral extends Expression {
  kind: "ArrayLiteral";
  elements: Expression[];
}

export interface ArrowFnExpr extends Expression {
  kind: "ArrowFnExpr";
  parameters: string[];
  body: Statement[] | Expression;
  async: boolean;
}

File: frontend/ast.go
package frontend

type NodeType string

const (
	ProgramNode              NodeType = "Program"
	VarDeclarationNode       NodeType = "VarDeclaration"
	GlobalDeclarationNode    NodeType = "GlobalDeclaration"
	FunctionDeclarationNode  NodeType = "FunctionDeclaration"
	IfStatementNode          NodeType = "IfStatement"
	WhileStatementNode       NodeType = "WhileStatement"
	ForStatementNode         NodeType = "ForStatement"
	SwitchStatementNode      NodeType = "SwitchStatement"
	CaseStatementNode        NodeType = "CaseStatement"
	TryCatchStatementNode    NodeType = "TryCatchStatement"
	ReturnStatementNode      NodeType = "ReturnStatement"
	ThrowStatementNode       NodeType = "ThrowStatement"
	BreakStatementNode       NodeType = "BreakStatement"
	ContinueStatementNode    NodeType = "ContinueStatement"
	ImportStatementNode      NodeType = "ImportStatement"
	ImportExprNode           NodeType = "ImportExpr"
	AssignmentExprNode       NodeType = "AssignmentExpr"
	BinaryExprNode           NodeType = "BinaryExpr"
	CallExprNode             NodeType = "CallExpr"
	MemberExprNode           NodeType = "MemberExpr"
	AwaitExprNode            NodeType = "AwaitExpr"
	IdentifierNode           NodeType = "Identifier"
	NumericLiteralNode       NodeType = "NumericLiteral"
	StringLiteralNode        NodeType = "StringLiteral"
	ObjectLiteralNode        NodeType = "ObjectLiteral"
	PropertyNode             NodeType = "Property"
	ArrayLiteralNode         NodeType = "ArrayLiteral"
	ArrowFnExprNode          NodeType = "ArrowFnExpr"
	ExportDeclarationNode    NodeType = "ExportDeclaration"
	NamedImportStatementNode NodeType = "NamedImportStatement"
)

type Stmt interface {
	Kind() NodeType
	GetLine() int
	GetColumn() int
	GetFile() string
	SetFile(file string)
	GetSource() string
	SetSource(source string)
}

type Expr interface {
	Stmt
	exprNode()
}

type BaseNode struct {
	Type   NodeType
	Line   int
	Column int
	File   string
	Source string
}

func (b *BaseNode) Kind() NodeType { return b.Type }
func (b *BaseNode) GetLine() int { return b.Line }
func (b *BaseNode) GetColumn() int { return b.Column }
func (b *BaseNode) GetFile() string { return b.File }
func (b *BaseNode) SetFile(file string) { b.File = file }
func (b *BaseNode) GetSource() string { return b.Source }
func (b *BaseNode) SetSource(source string) { b.Source = source }

type Program struct {
	BaseNode
	Body []Stmt
}

type VarDeclaration struct {
	BaseNode
	Constant   bool
	Identifier string
	Value      Expr
}

type GlobalDeclaration struct {
	BaseNode
	Constant   bool
	Identifier string
	Value      Expr
}

type FunctionDeclaration struct {
	BaseNode
	Name       string
	Parameters []string
	Body       []Stmt
	Async      bool
}

type IfStatement struct {
	BaseNode
	Condition  Expr
	ThenBranch []Stmt
	ElseBranch []Stmt
}

type WhileStatement struct {
	BaseNode
	Condition Expr
	Body      []Stmt
}

type ForStatement struct {
	BaseNode
	Counter   string
	Start     Expr
	End       Expr
	Init      Stmt
	Condition Expr
	Update    Expr
	Body      []Stmt
}

type SwitchStatement struct {
	BaseNode
	Discriminant Expr
	Cases        []CaseStatement
	Default      []Stmt
}

type CaseStatement struct {
	BaseNode
	Test       Expr
	Consequent []Stmt
}

type TryCatchStatement struct {
	BaseNode
	Body           []Stmt
	CatchParameter string
	CatchBlock     []Stmt
	FinallyBlock   []Stmt
}

type ReturnStatement struct {
	BaseNode
	Value Expr
}

type ThrowStatement struct {
	BaseNode
	Argument Expr
}

type BreakStatement struct {
	BaseNode
}

type ContinueStatement struct {
	BaseNode
}

type ImportStatement struct {
	BaseNode
	ModuleName string
}

type NamedImportStatement struct {
	BaseNode
	Imports    []string
	ModuleName string
}

type ExportDeclaration struct {
	BaseNode
	Declaration Stmt
}

// Expressions
func (b *BaseNode) exprNode() {}

type ImportExpr struct {
	BaseNode
	ModuleName Expr
}

type AssignmentExpr struct {
	BaseNode
	Assignee Expr
	Value    Expr
}

type BinaryExpr struct {
	BaseNode
	Left     Expr
	Right    Expr
	Operator string
}

type CallExpr struct {
	BaseNode
	Args   []Expr
	Caller Expr
}

type MemberExpr struct {
	BaseNode
	Object   Expr
	Property Expr
	Computed bool
	Optional bool
}

type AwaitExpr struct {
	BaseNode
	Argument Expr
}

type Identifier struct {
	BaseNode
	Symbol string
}

type NumericLiteral struct {
	BaseNode
	Value float64
}

type StringLiteral struct {
	BaseNode
	Value string
}

type Property struct {
	BaseNode
	Key   string
	Value Expr
}

type ObjectLiteral struct {
	BaseNode
	Properties []Property
}

type ArrayLiteral struct {
	BaseNode
	Elements []Expr
}

type ArrowFnExpr struct {
	BaseNode
	Parameters []string
	Body       []Stmt
	Async      bool
}
